lib/onelogin/ruby-saml/metadata.rb

```diff
@@ -32,21 +32,20 @@ def generate(settings, pretty_print=false)
             "WantAssertionsSigned" => settings.security[:want_assertions_signed],
         }
 
-        # Add KeyDescriptor if messages will be signed / encrypted
         cert = settings.get_sp_cert
         if cert
-          cert_text = Base64.encode64(cert.to_der).gsub("\n", '')
-          kd = sp_sso.add_element "md:KeyDescriptor", { "use" => "signing" }
-          ki = kd.add_element "ds:KeyInfo", {"xmlns:ds" => "http://www.w3.org/2000/09/xmldsig#"}
-          xd = ki.add_element "ds:X509Data"
-          xc = xd.add_element "ds:X509Certificate"
-          xc.text = cert_text
-
-          kd2 = sp_sso.add_element "md:KeyDescriptor", { "use" => "encryption" }
-          ki2 = kd2.add_element "ds:KeyInfo", {"xmlns:ds" => "http://www.w3.org/2000/09/xmldsig#"}
-          xd2 = ki2.add_element "ds:X509Data"
-          xc2 = xd2.add_element "ds:X509Certificate"
-          xc2.text = cert_text
+          sp_cert_text = get_cert_text(cert)
+
+          add_sp_cert(sp_sso, "signing", sp_cert_text)
+          add_sp_cert(sp_sso, "encryption", sp_cert_text)
+
+          intermediate_certs = settings.get_sp_intermediate_certs
+          if intermediate_certs
+            intermediate_certs.each do |intermediate_cert|
+              intermediate_cert_text = get_cert_text(intermediate_cert)
+              add_sp_cert(sp_sso, "encryption", intermediate_cert_text)
```

Comment:
In other words, `intermediate_cert_texts` should be an additional option to the two calls to `add_sp_cert` above.